### ISSUE TITLE
Making my own model from KPConv layers

### ISSUE THREAD

**Issue by IdeGelis** [nicolas-chaulet/torch-points3d#539]
Hello all, 

First of all, thank you very much for your interesting and valuable work.

I would like to test an idea for a network in order to process 3D ALS data. I would like to implement a new architecture where some layers contain kernel point convolutions in order to solve others tasks than segmentation and classification.

For now, I see two solutions:
- First, using TorchPoint 3D : but I need to modify the UNet architecture. I saw in other issues that I can modify the yaml file ([https://github.com/nicolas-chaulet/torch-points3d/tree/master/torch_points3d/applications/conf/kpconv](https://github.com/nicolas-chaulet/torch-points3d/tree/master/torch_points3d/applications/conf/kpconv)) but I am not sure how to do it.
I wonder where does the yaml is read? Can I create whatever architecture I want or should I respect the variable name 'DownConv' and just modify the number of layer (which is not what I want) ? I am not used to yaml files. I also had a look at the notebook ([https://github.com/nicolas-chaulet/torch-points3d/blob/master/notebooks/PartSegmentationKPConv.ipynb](https://github.com/nicolas-chaulet/torch-points3d/blob/master/notebooks/PartSegmentationKPConv.ipynb)), but I am not sure that I can modify as I want the forward part of the PartSegKPConv class. 
- Secondly, using KPConv implementation of H. Thomas and modify the model in Pytorch. However, I feel that using TorchPoint3D can be easier for example for the part of loading data. Indeed pre-implemented classes of network could possibly make the implementation easier...

To summarize my questions: do you think using TorchPoint3D in order to define a new architecture which use some existing KPConv blocks is a good idea? When I define the forward part of the model class can I do it like in pytorch, for example using a KPConv block defined in a Yaml file (or directly imported if there is an existing KPConv layer) and then use normal pytorch operator between tensors to manipulate deeper features? 

Furthermore, in the implementation of H. Thomas of KPConv, they do a grid subsampling. Is there an existing grid subsampling function implemented that I can use on a new dataset?
Finally, I didn't find where is done the kernel point selection.

Thank you so much in advance for your help!

Best regards,

Iris

**Comment by nicolas-chaulet** [comment]
Hi,
Yes tp3d is a great place for, we built it for that reason. If you want to implement your own architecture you can just do that as usual, just create a new model in the models folder and a yaml file that refers to it (see [here](https://torch-points3d.readthedocs.io/en/latest/src/tutorials.html#create-a-new-model) for the doc). You can keep the yaml very minimal and include all the specific configs in code, like it is done for Minkowski for example:
https://github.com/nicolas-chaulet/torch-points3d/blob/master/torch_points3d/models/segmentation/minkowski.py
with the config file here:
https://github.com/nicolas-chaulet/torch-points3d/blob/master/conf/models/segmentation/minkowski_baseline.yaml
I hope this helps!

**Comment by IdeGelis** [comment]
Hi,
Thank you very much for your reply. I am trying to implement it, I will come back to you if I have any question!
Regards!
Iris

**Comment by IdeGelis** [comment]
Hi @nicolas-chaulet , 
I began to develop my model based on KPConv and I have an error which is rising when I launch a training. This error occurs during the encoder part of KPConv which I did not modify.... I also tried to do a segmentation task on my dataset that I implemented with the Unet KPConv model in the segmentation folder, and it returned the same error. I tested on my dataset, so maybe it comes from my dataset definition but I don't understand why. My dataset seems to pass the first part of preparation and initialization of the model, it is during training that I get the error... 

If you have an idea of what is going on I would be grateful!
Thank you,
Iris

Here you have what has been returned when I launch the train.py file in torch-points3d on the segmentation task on my dataset with the Unet KPConv model: 

> DEVICE : cuda
Setting batchnorm momentum at 0.1
KPConvPaper(
  (down_modules): ModuleList(
    (0): KPDualBlock(
      Nb parameters: 31424
      (blocks): ModuleList(
        (0): SimpleBlock(
          Nb parameters: 1088; None; RadiusNeighbourFinder {'_radius': 0.125, '_max_num_neighbors': 25, '_conv_type': 'partial_dense'}
          (kp_conv): KPConvLayer(InF: 1, OutF: 64, kernel_pts: 15, radius: 0.08, KP_influence: linear, Add_one: False)
          (bn): FastBatchNorm1d(
            (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          )
          (activation): LeakyReLU(negative_slope=0.1)
        )
        (1): ResnetBBlock(
          Nb parameters: 30336
          (kp_conv): SimpleBlock(
            Nb parameters: 15424; None; RadiusNeighbourFinder {'_radius': 0.125, '_max_num_neighbors': 25, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 32, OutF: 32, kernel_pts: 15, radius: 0.08, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=64, out_features=32, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=32, out_features=128, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Sequential(
            (0): Linear(in_features=64, out_features=128, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
          )
          (activation): LeakyReLU(negative_slope=0.1)
        )
      )
    )
    (1): KPDualBlock(
      Nb parameters: 144000
      (blocks): ModuleList(
        (0): ResnetBBlock(
          Nb parameters: 23936
          (kp_conv): SimpleBlock(
            Nb parameters: 15424; GridSampling3D(grid_size=0.1, quantize_coords=False, mode=mean); RadiusNeighbourFinder {'_radius': 0.125, '_max_num_neighbors': 25, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 32, OutF: 32, kernel_pts: 15, radius: 0.08, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=128, out_features=32, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=32, out_features=128, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Identity()
          (activation): LeakyReLU(negative_slope=0.1)
        )
        (1): ResnetBBlock(
          Nb parameters: 120064
          (kp_conv): SimpleBlock(
            Nb parameters: 61568; None; RadiusNeighbourFinder {'_radius': 0.25, '_max_num_neighbors': 30, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 64, OutF: 64, kernel_pts: 15, radius: 0.15, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=128, out_features=64, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=64, out_features=256, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Sequential(
            (0): Linear(in_features=128, out_features=256, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
          )
          (activation): LeakyReLU(negative_slope=0.1)
        )
      )
    )
    (2): KPDualBlock(
      Nb parameters: 572672
      (blocks): ModuleList(
        (0): ResnetBBlock(
          Nb parameters: 94976
          (kp_conv): SimpleBlock(
            Nb parameters: 61568; GridSampling3D(grid_size=0.2, quantize_coords=False, mode=mean); RadiusNeighbourFinder {'_radius': 0.25, '_max_num_neighbors': 30, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 64, OutF: 64, kernel_pts: 15, radius: 0.15, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=256, out_features=64, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=64, out_features=256, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Identity()
          (activation): LeakyReLU(negative_slope=0.1)
        )
        (1): ResnetBBlock(
          Nb parameters: 477696
          (kp_conv): SimpleBlock(
            Nb parameters: 246016; None; RadiusNeighbourFinder {'_radius': 0.5, '_max_num_neighbors': 38, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 128, OutF: 128, kernel_pts: 15, radius: 0.30, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=256, out_features=128, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=128, out_features=512, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Sequential(
            (0): Linear(in_features=256, out_features=512, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
          )
          (activation): LeakyReLU(negative_slope=0.1)
        )
      )
    )
    (3): KPDualBlock(
      Nb parameters: 2284032
      (blocks): ModuleList(
        (0): ResnetBBlock(
          Nb parameters: 378368
          (kp_conv): SimpleBlock(
            Nb parameters: 246016; GridSampling3D(grid_size=0.4, quantize_coords=False, mode=mean); RadiusNeighbourFinder {'_radius': 0.5, '_max_num_neighbors': 38, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 128, OutF: 128, kernel_pts: 15, radius: 0.30, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=512, out_features=128, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=128, out_features=512, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Identity()
          (activation): LeakyReLU(negative_slope=0.1)
        )
        (1): ResnetBBlock(
          Nb parameters: 1905664
          (kp_conv): SimpleBlock(
            Nb parameters: 983552; None; RadiusNeighbourFinder {'_radius': 1.0, '_max_num_neighbors': 38, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 256, OutF: 256, kernel_pts: 15, radius: 0.60, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=512, out_features=256, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=256, out_features=1024, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Sequential(
            (0): Linear(in_features=512, out_features=1024, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
          )
          (activation): LeakyReLU(negative_slope=0.1)
        )
      )
    )
    (4): KPDualBlock(
      Nb parameters: 9122816
      (blocks): ModuleList(
        (0): ResnetBBlock(
          Nb parameters: 1510400
          (kp_conv): SimpleBlock(
            Nb parameters: 983552; GridSampling3D(grid_size=0.8, quantize_coords=False, mode=mean); RadiusNeighbourFinder {'_radius': 1.0, '_max_num_neighbors': 38, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 256, OutF: 256, kernel_pts: 15, radius: 0.60, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=1024, out_features=256, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=256, out_features=1024, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Identity()
          (activation): LeakyReLU(negative_slope=0.1)
        )
        (1): ResnetBBlock(
          Nb parameters: 7612416
          (kp_conv): SimpleBlock(
            Nb parameters: 3933184; None; RadiusNeighbourFinder {'_radius': 2.0, '_max_num_neighbors': 38, '_conv_type': 'partial_dense'}
            (kp_conv): KPConvLayer(InF: 512, OutF: 512, kernel_pts: 15, radius: 1.20, KP_influence: linear, Add_one: False)
            (bn): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (activation): LeakyReLU(negative_slope=0.1)
          )
          (unary_1): Sequential(
            (0): Linear(in_features=1024, out_features=512, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (unary_2): Sequential(
            (0): Linear(in_features=512, out_features=2048, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
            (2): LeakyReLU(negative_slope=0.1)
          )
          (shortcut_op): Sequential(
            (0): Linear(in_features=1024, out_features=2048, bias=False)
            (1): FastBatchNorm1d(
              (batch_norm): BatchNorm1d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
            )
          )
          (activation): LeakyReLU(negative_slope=0.1)
        )
      )
    )
  )
  (inner_modules): ModuleList(
    (0): Identity()
  )
  (up_modules): ModuleList(
    (0): FPModule_PD(
      (nn): Sequential(
        (0): Sequential(
          (0): Linear(in_features=3072, out_features=512, bias=False)
          (1): FastBatchNorm1d(
            (batch_norm): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          )
          (2): LeakyReLU(negative_slope=0.2)
        )
      )
    )
    (1): FPModule_PD(
      (nn): Sequential(
        (0): Sequential(
          (0): Linear(in_features=1024, out_features=256, bias=False)
          (1): FastBatchNorm1d(
            (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          )
          (2): LeakyReLU(negative_slope=0.2)
        )
      )
    )
    (2): FPModule_PD(
      (nn): Sequential(
        (0): Sequential(
          (0): Linear(in_features=512, out_features=128, bias=False)
          (1): FastBatchNorm1d(
            (batch_norm): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          )
          (2): LeakyReLU(negative_slope=0.2)
        )
      )
    )
    (3): FPModule_PD(
      (nn): Sequential(
        (0): Sequential(
          (0): Linear(in_features=256, out_features=64, bias=False)
          (1): FastBatchNorm1d(
            (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          )
          (2): LeakyReLU(negative_slope=0.2)
        )
      )
    )
  )
  (FC_layer): Sequential(
    (1): Sequential(
      (0): Linear(in_features=64, out_features=64, bias=False)
      (1): FastBatchNorm1d(
        (batch_norm): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (2): LeakyReLU(negative_slope=0.2)
    )
    (Class): Linear(in_features=64, out_features=3, bias=False)
    (Softmax): LogSoftmax(dim=-1)
  )
)
Optimizer: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    initial_lr: 0.001
    lr: 0.001
    weight_decay: 0
)
Learning Rate Scheduler: ExponentialLR({'gamma': 0.9885}, update_scheduler_on=on_epoch)
BatchNorm Scheduler: BNMomentumScheduler(base_momentum: 0.1, update_scheduler_on=on_epoch)
Accumulated gradients: None
Model size = 14078208
Dataset: Urb3DSimulDataset 
train_pre_batch_collate_transform = None
val_pre_batch_collate_transform = None
test_pre_batch_collate_transform = None
pre_transform = None
test_transform = None
train_transform = None
val_transform = None
inference_transform = None
Size of train_dataset = 14901
Size of test_dataset = 8359
Size of val_dataset = 10338
Batch size = 16
Available stage selection datasets:  ['test', 'val'] 
The models will be selected using the metrics on following dataset:  val 
Access tensorboard with the following command <tensorboard --logdir=/share/home/degelis/dev/deep3DT/torch-points3d/outputs/firstTest/firstTest-KPConvPaper-20210127_175142/tensorboard>
EPOCH 1 / 100
  0%|                                                                            | 0/932 [00:11<?, ?it/s]
Traceback (most recent call last):
  File "train_Urb3DSimul.py", line 20, in <module>
    main()
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/main.py", line 24, in decorated_main
    strict=strict,
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/_internal/utils.py", line 174, in run_hydra
    overrides=args.overrides,
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/_internal/hydra.py", line 86, in run
    job_subdir_key=None,
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/plugins/common/utils.py", line 109, in run_job
    ret.return_value = task_function(task_cfg)
  File "train_Urb3DSimul.py", line 12, in main
    trainer.train()
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/trainer_Urb3DS.py", line 138, in train
    self._train_epoch(epoch)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/trainer_Urb3DS.py", line 191, in _train_epoch
    self._model.optimize_parameters(epoch, self._dataset.batch_size)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/models/base_model.py", line 215, in optimize_parameters
    self.forward(epoch=epoch)  # first call forward to calculate intermediate results
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/models/segmentation/kpconv.py", line 112, in forward
    data = self.down_modules[i](data, precomputed=self.pre_computed)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/blocks.py", line 280, in forward
    data = block(data, precomputed=precomputed)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/blocks.py", line 201, in forward
    output = self.kp_conv(output, precomputed=precomputed)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/blocks.py", line 87, in forward
    x = self.kp_conv(q_pos, data.pos, idx_neighboors, data.x,)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/kernels.py", line 92, in forward
    self.aggregation_mode,
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/convolution_ops.py", line 53, in KPConv_ops
    neighbors = gather(support_points, neighbors_indices)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/core/common_modules/gathering.py", line 10, in gather
    idx[idx == -1] = x.shape[0] - 1  # Shadow point
RuntimeError: CUDA error: device-side assert triggered
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:115: operator(): block: [18,0,0], thread: [2,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:115: operator(): block: [15,0,0], thread: [37,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:115: operator(): block: [15,0,0], thread: [17,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:115: operator(): block: [14,0,0], thread: [4,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.

**Comment by IdeGelis** [comment]
Hi, 
I am updating my last issue because I saw on forums that whithout setting the variable CUDA_LAUNCH_BLOCKING at 1, the displayed error could not be the effective one. 
So I added `import os
os.environ['CUDA_LAUNCH_BLOCKING'] = "1"` in my code, and now I have the following error when it begins to train, but I still don't understand the error. 
>   0%|                                                                            | 0/932 [00:01<?, ?it/s]
Traceback (most recent call last):
  File "train.py", line 22, in <module>
    main()
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/main.py", line 24, in decorated_main
    strict=strict,
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/_internal/utils.py", line 174, in run_hydra
    overrides=args.overrides,
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/_internal/hydra.py", line 86, in run
    job_subdir_key=None,
  File "/share/home/degelis/.local/lib/python3.7/site-packages/hydra/plugins/common/utils.py", line 109, in run_job
    ret.return_value = task_function(task_cfg)
  File "train.py", line 14, in main
    trainer.train()
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/trainer.py", line 140, in train
    self._train_epoch(epoch)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/trainer.py", line 194, in _train_epoch
    self._model.optimize_parameters(epoch, self._dataset.batch_size)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/models/base_model.py", line 215, in optimize_parameters
    self.forward(epoch=epoch)  # first call forward to calculate intermediate results
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/models/segmentation/kpconv.py", line 112, in forward
    data = self.down_modules[i](data, precomputed=self.pre_computed)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/blocks.py", line 280, in forward
    data = block(data, precomputed=precomputed)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/blocks.py", line 201, in forward
    output = self.kp_conv(output, precomputed=precomputed)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/modules/KPConv/blocks.py", line 84, in forward
    idx_neighboors = self.neighbour_finder(data.pos, q_pos, batch_x=data.batch, batch_y=q_batch)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/core/spatial_ops/neighbour_finder.py", line 15, in __call__
    return self.find_neighbours(x, y, batch_x, batch_y)
  File "/share/home/degelis/dev/deep3DT/torch-points3d/torch_points3d/core/spatial_ops/neighbour_finder.py", line 36, in find_neighbours
    self._radius, self._max_num_neighbors, x, y, mode=self._conv_type, batch_x=batch_x, batch_y=batch_y
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch_points_kernels/torchpoints.py", line 210, in ball_query
    return ball_query_partial_dense(radius, nsample, x, y, batch_x, batch_y, sort=sort)
  File "/share/home/degelis/.local/lib/python3.7/site-packages/torch_points_kernels/torchpoints.py", line 165, in ball_query_partial_dense
    ind, dist = tpcuda.ball_query_partial_dense(x, y, batch_x, batch_y, radius, nsample)
RuntimeError: CUDA error: device-side assert triggered

Moreover I tested with Shapenet dataset to compare to my data,  I saw that when I print the data given to the model for the forward step, data are like that `Batch(batch=[10240], category=[10240], grid_size=[5], id_scan=[5], origin_id=[10240], pos=[10240, 3], x=[10240, 4], y=[10240])` and mine are `Batch(batch=[6820], pos=[6820, 3], x=[6820, 1], y=[6820])
`. Dimensions of pos, x and y seem to be coherent. I don't know what I am missing...

Thank you in advance for your help,
Iris

**Comment by nicolas-chaulet** [comment]
Hi Iris, this could be related to a degenerate case where there is only on point at a given scale for example. A good place to start debugging would be to set the batch size to 1, run on CPU and have a break point in your code so that you can see what happens.

**Comment by IdeGelis** [comment]
Hi Nicolas,
Thank you for your answer, indeed, my point clouds were too small a certain scale. 
Thank you very much,
Iris
